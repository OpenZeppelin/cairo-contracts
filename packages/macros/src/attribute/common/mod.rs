pub mod text_span;
