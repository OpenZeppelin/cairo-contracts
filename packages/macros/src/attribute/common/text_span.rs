use std::collections::{HashMap, VecDeque};

use cairo_lang_macro::{quote, TextSpan, Token, TokenStream, TokenTree};
use cairo_lang_parser::utils::SimpleParserDatabase;
use cairo_lang_syntax::node::with_db::SyntaxNodeWithDb;

fn make_ident(text: &str, span: TextSpan) -> TokenTree {
    TokenTree::Ident(Token::new(text, span))
}

fn tokenize_str(s: &str, db: &SimpleParserDatabase) -> TokenStream {
    let syntax_node = db.parse_virtual(s).unwrap();
    let syntax_node_with_db = SyntaxNodeWithDb::new(&syntax_node, db);
    quote! {#syntax_node_with_db}
}

/// Merge spans from `initial_tokens` into a tokenized version of `final_output`.
/// - If a token text in `final_output` matches one from `initial_tokens` (by exact text),
///   it inherits that tokenâ€™s span (first-come, first-served).
/// - Otherwise, it gets `TextSpan::call_site()`.
pub fn merge_spans_from_initial(
    initial_tokens: &[TokenTree],
    final_output: &str,
    db: &SimpleParserDatabase,
) -> Vec<TokenTree> {
    // Build a multimap: text -> queue of spans (to support duplicates, left-to-right)
    let mut span_index: HashMap<String, VecDeque<TextSpan>> = HashMap::new();
    for tt in initial_tokens {
        let TokenTree::Ident(tok) = tt;
        let text = tok.content.to_string();
        span_index
            .entry(text)
            .or_default()
            .push_back(tok.span.clone());
    }

    // Tokenize final output
    let final_tokens = tokenize_str(final_output, db)
        .tokens
        .iter()
        .map(|tt| {
            let TokenTree::Ident(tok) = tt;
            tok.content.to_string()
        })
        .collect::<Vec<_>>();

    // Rebuild final tokens, reusing spans when the text matches; else call_site.
    let mut out = Vec::with_capacity(final_tokens.len());
    for text in final_tokens {
        if let Some(queue) = span_index.get_mut(&text) {
            if let Some(span) = queue.pop_front() {
                out.push(make_ident(&text, span));
                if queue.is_empty() {
                    // keep the map clean
                    span_index.remove(&text);
                }
                continue;
            }
        }
        out.push(make_ident(&text, TextSpan::call_site()));
    }

    out
}
