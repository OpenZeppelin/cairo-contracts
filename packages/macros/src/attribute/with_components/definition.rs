use cairo_lang_formatter::format_string;
use cairo_lang_macro::{attribute_macro, Diagnostic, ProcMacroResult, TokenStream};
use cairo_lang_parser::utils::SimpleParserDatabase;
use regex::Regex;

use super::{components::AllowedComponents, parser::WithComponentsParser};

/// Inserts multiple component dependencies into a modules codebase.
#[attribute_macro]
pub fn with_components(attribute_stream: TokenStream, item_stream: TokenStream) -> ProcMacroResult {
    let args = parse_args(&attribute_stream.to_string());

    // 1. Get the components info (if valid)
    let mut components_info = vec![];
    let empty_result = ProcMacroResult::new(TokenStream::empty());
    for arg in args {
        let maybe_component = AllowedComponents::from_str(&arg);
        match maybe_component {
            Ok(component) => {
                components_info.push(component.get_info());
            }
            Err(err) => {
                return empty_result.with_diagnostics(err.into());
            }
        }
    }

    // 2. Parse the item stream
    let db = SimpleParserDatabase::default();
    let (content, diagnostics) = match db.parse_virtual(item_stream.to_string()) {
        Ok(node) => WithComponentsParser::new(node, &components_info).parse(&db),
        Err(err) => {
            let error = Diagnostic::error(err.format(&db));
            return empty_result.with_diagnostics(error.into());
        }
    };

    let formatted_content = if !content.is_empty() {
        format_string(&db, content)
    } else {
        content
    };

    ProcMacroResult::new(TokenStream::new(formatted_content)).with_diagnostics(diagnostics)
}

/// Parses the arguments from the attribute stream.
fn parse_args(text: &str) -> Vec<String> {
    let re = Regex::new(r"(\w+)").unwrap();
    let matches = re.find_iter(text);
    matches.map(|m| m.as_str().to_string()).collect()
}

#[cfg(test)]
mod tests {
    use super::parse_args;

    #[test]
    fn test_parse_args() {
        let attribute = "(ERC20, Ownable)";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["ERC20", "Ownable"]);

        let attribute = "ERC20";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["ERC20"]);

        let attribute = "(Ownable, ERC20, Other, Another)";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["Ownable", "ERC20", "Other", "Another"]);
    }
}
