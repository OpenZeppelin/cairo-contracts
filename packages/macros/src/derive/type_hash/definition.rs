use cairo_lang_formatter::format_string;
use cairo_lang_macro::{attribute_macro, Diagnostic, Diagnostics, ProcMacroResult, TokenStream};
use cairo_lang_parser::utils::SimpleParserDatabase;
use cairo_lang_plugins::plugins::utils::PluginTypeInfo;
use cairo_lang_starknet_classes::keccak::starknet_keccak;
use cairo_lang_syntax::node::{ast, TypedSyntaxNode};
use cairo_lang_syntax::node::{db::SyntaxGroup, SyntaxNode};
use convert_case::{Case, Casing};
use indoc::formatdoc;
use regex::Regex;

use crate::type_hash::parser::TypeHashParser;

use super::diagnostics::errors;
use super::parser::parse_string_arg;

/// Derive macro that generates a SNIP-12 type hash constant for a struct.
///
/// Example:
/// ```
/// #[type_hash]
/// pub struct MyStruct {
///     pub some_member: felt252,
/// }
///
/// // Generates:
/// pub const MY_STRUCT_TYPE_HASH: felt252 = 0x[HASH];
/// ```
#[attribute_macro]
pub fn type_hash(attr_stream: TokenStream, item_stream: TokenStream) -> ProcMacroResult {
    // 1. Parse the attribute stream
    let config = match parse_args(&attr_stream.to_string()) {
        Ok(config) => config,
        Err(err) => {
            return ProcMacroResult::new(TokenStream::empty()).with_diagnostics(err.into());
        }
    };

    // 2. Parse the item stream
    let db = SimpleParserDatabase::default();
    let content = match db.parse_virtual(item_stream.to_string()) {
        Ok(node) => handle_node(&db, node, &config),
        Err(err) => {
            let error = Diagnostic::error(err.format(&db));
            return ProcMacroResult::new(TokenStream::empty()).with_diagnostics(error.into());
        }
    };

    // 3. Format the expanded content
    let (formatted_content, diagnostics) = match content {
        Ok(content) => (format_string(&db, content), Diagnostics::new(vec![])),
        Err(err) => (String::new(), err.into()),
    };

    // 4. Return the result
    let result = TokenStream::new(item_stream.to_string() + &formatted_content);
    ProcMacroResult::new(result).with_diagnostics(diagnostics)
}

/// This attribute macro is used to specify an override for the SNIP-12 type.
///
/// It doesn't modify the source code directly, but it is used in the type hash parser to generate the new type hash.
///
/// Example:
/// ```
/// #[type_hash]
/// pub struct MyStruct {
///     #[snip12(name: "Some Member", kind: "shortstring")]
///     pub some_member: felt252,
/// }
/// ```
#[attribute_macro]
pub fn snip12(attr_stream: TokenStream, item_stream: TokenStream) -> ProcMacroResult {
    // Validate the received format
    let re1 = Regex::new(r"^\(name: (.*), kind: (.*)\)$").unwrap();
    let re2 = Regex::new(r"^\(kind: (.*), name: (.*)\)$").unwrap();
    let re3 = Regex::new(r"^\(kind: (.*)\)$").unwrap();
    let re4 = Regex::new(r"^\(name: (.*)\)$").unwrap();
    let s = attr_stream.to_string();

    if re1.is_match(&s) || re2.is_match(&s) || re3.is_match(&s) || re4.is_match(&s) {
        ProcMacroResult::new(item_stream)
    } else {
        let error = Diagnostic::error(errors::INVALID_SNIP12_ATTRIBUTE_FORMAT);
        ProcMacroResult::new(item_stream).with_diagnostics(error.into())
    }
}

/// Configuration for the type hash attribute.
///
/// Represents the arguments passed to the type_hash attribute.
///
/// Example:
/// ```
/// #[type_hash(name: "MyStruct", debug: true)]
/// ```
pub struct TypeHashArgs {
    pub name: String,
    pub debug: bool,
}

/// Parses the arguments passed to the type_hash attribute and
/// returns a TypeHashArgs struct containing the parsed arguments.
fn parse_args(s: &str) -> Result<TypeHashArgs, Diagnostic> {
    let mut args = TypeHashArgs {
        name: String::new(),
        debug: false,
    };
    // If the attribute is empty, return the default config
    if s.is_empty() || s == "()" {
        return Ok(args);
    }

    let allowed_args = ["name", "debug"];
    let allowed_args_re = allowed_args.join("|");

    let re = Regex::new(&format!(
        r#"^\(({}): ([\w"' ])+(?:, ({}): ([\w"' ])+)*\)$"#,
        allowed_args_re, allowed_args_re
    ))
    .unwrap();

    if re.is_match(s) {
        // Remove the parentheses
        let s = &s[1..s.len() - 1];
        for arg in s.split(",") {
            let parts = arg.split(":").collect::<Vec<&str>>();
            let name = parts[0].trim();
            let value = parts[1].trim();
            match name {
                "name" => args.name = parse_string_arg(value)?,
                "debug" => args.debug = value == "true",
                // This should be unreachable as long as the regex is correct
                _ => panic!("Invalid argument: {}", name),
            }
        }
        Ok(args)
    } else {
        Err(Diagnostic::error(
            errors::INVALID_TYPE_HASH_ATTRIBUTE_FORMAT,
        ))
    }
}

fn handle_node(
    db: &dyn SyntaxGroup,
    node: SyntaxNode,
    args: &TypeHashArgs,
) -> Result<String, Diagnostic> {
    let typed = ast::SyntaxFile::from_syntax_node(db, node);
    let items = typed.items(db).elements(db);

    let Some(item_ast) = items.first() else {
        let error = Diagnostic::error(errors::EMPTY_TYPE_FOUND);
        return Err(error);
    };

    // Generate type hash for structs/enums only
    match item_ast {
        ast::ModuleItem::Struct(_) | ast::ModuleItem::Enum(_) => {
            // It is safe to unwrap here because we know the item is a struct
            let plugin_type_info = PluginTypeInfo::new(db, item_ast).unwrap();
            generate_code(db, &plugin_type_info, args)
        }
        _ => {
            let error = Diagnostic::error(errors::NOT_VALID_TYPE_TO_DECORATE);
            Err(error)
        }
    }
}

/// Generates the code for the type hash constant.
fn generate_code(
    db: &dyn SyntaxGroup,
    plugin_type_info: &PluginTypeInfo,
    args: &TypeHashArgs,
) -> Result<String, Diagnostic> {
    let mut parser = TypeHashParser::new(plugin_type_info);
    let type_hash_string = parser.parse(db, args)?;
    let type_hash = starknet_keccak(type_hash_string.as_bytes());
    let type_name = plugin_type_info.name.as_str().to_case(Case::UpperSnake);

    let debug_string = if args.debug {
        formatdoc!(
            r#"
            pub fn __{type_name}_encoded_type() {{
                println!("{}");
            }}"#,
            type_hash_string.replace("\"", "\\\"")
        )
    } else {
        String::new()
    };

    let code = formatdoc!(
        "
        {debug_string}
        pub const {type_name}_TYPE_HASH: felt252 = 0x{:x};

        ",
        type_hash
    );

    Ok(code)
}
