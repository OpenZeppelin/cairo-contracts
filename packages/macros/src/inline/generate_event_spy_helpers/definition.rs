use cairo_lang_formatter::format_string;
use cairo_lang_macro::{inline_macro, Diagnostic, ProcMacroResult, TokenStream};
use cairo_lang_parser::utils::SimpleParserDatabase;
use proc_macro2::TokenStream as ProcTokenStream;
use quote::{format_ident, quote};

use crate::{generate_event_spy_helpers::parser, utils::camel_to_snake};

/// Generates helper functions for spying on events in tests.
///
/// Example:
/// ```
/// generate_event_spy_helpers! {
///     impl AccessControlDefaultAdminRulesSpyHelpers {
///         event DefaultAdminTransferScheduled(
///            #[key]
///            new_admin: ContractAddress,
///            accept_schedule: u64
///         );
///     }
/// }
/// ```
///
/// // Generated code:
/// #[generate_trait]
/// impl AccessControlDefaultAdminRulesSpyHelpers of AccessControlDefaultAdminRulesSpyHelpersTrait {
///     fn assert_event_default_admin_transfer_scheduled(
///         ref self: EventSpy,
///         contract: ContractAddress,
///         new_admin: ContractAddress,
///         accept_schedule: u64,
///     ) {
///         let mut keys = array![];
///         keys.append_serde(selector!("DefaultAdminTransferScheduled"));
///         keys.append_serde(new_admin);
///
///         let mut data = array![];
///         data.append_serde(accept_schedule);
///
///         let expected = Event { keys, data };
///         self.assert_only_event(contract, expected);
///     }
///
///     fn assert_only_event_default_admin_transfer_scheduled(
///         ref self: EventSpy,
///         contract: ContractAddress,
///         new_admin: ContractAddress,
///         accept_schedule: u64,
///     ) {
///         self.assert_event_default_admin_transfer_scheduled(contract, new_admin, accept_schedule);
///         self.assert_no_events_left_from(contract);
///     }
/// }
/// ```
#[inline_macro]
pub fn generate_event_spy_helpers(token_stream: TokenStream) -> ProcMacroResult {
    // Parse the arguments
    let impl_block = match parser::parse_dsl(&token_stream.to_string()) {
        Ok((_, impl_block)) => impl_block,
        Err(e) => {
            let error = Diagnostic::error(e.to_string());
            let empty_result = ProcMacroResult::new(TokenStream::empty());
            return empty_result.with_diagnostics(error.into());
        }
    };

    // Generate the helper functions
    let expanded = generate_code(&impl_block).to_string();

    // Format the code
    let db = SimpleParserDatabase::default();
    let formatted_code = format_string(&db, expanded);

    // Return the result
    ProcMacroResult::new(TokenStream::new(formatted_code))
}

/// Generates the code for event spy helper functions based on the provided implementation block.
///
/// # Arguments
///
/// * `impl_block` - The parsed implementation block containing event definitions
///
/// # Returns
///
/// A `ProcTokenStream` containing the generated code for event spy helper functions
fn generate_code(impl_block: &parser::ImplBlock) -> ProcTokenStream {
    let impl_name = format_ident!("{}", impl_block.name);
    let trait_name = format_ident!("{}Trait", impl_block.name);

    let methods: Vec<ProcTokenStream> = impl_block
        .events
        .iter()
        .map(|event| {
            let event_name = &event.name;
            let fn_suffix = format_ident!("{}", camel_to_snake(event_name));
            let selector = event_name.to_string();

            let mut key_fields = vec![];
            let mut data_fields = vec![];

            let mut fn_args = vec![];
            for param in &event.fields {
                let name = format_ident!("{}", param.name);
                let ty = format_ident!("{}", param.ty);
                fn_args.push(quote! { #name: #ty });

                if param.is_key {
                    key_fields.push(quote! { keys.append_serde(#name); });
                } else {
                    data_fields.push(quote! { data.append_serde(#name); });
                }
            }

            let assert_fn_name = format_ident!("assert_event_{}", fn_suffix);
            let only_fn_name = format_ident!("assert_only_event_{}", fn_suffix);
            let fields = event
                .fields
                .iter()
                .map(|p| format_ident!("{}", p.name))
                .collect::<Vec<_>>();

            let assert_only_fn = if event.is_only {
                quote! {
                    fn #only_fn_name(
                        ref self: EventSpy,
                        contract: ContractAddress,
                        #(#fn_args),*
                    ) {
                        self.#assert_fn_name(contract, #(#fields),*);
                        self.assert_no_events_left_from(contract);
                    }
                }
            } else {
                quote! {}
            };

            quote! {
                fn #assert_fn_name(
                    ref self: EventSpy,
                    contract: ContractAddress,
                    #(#fn_args),*
                ) {
                    let mut keys = array![];
                    keys.append_serde(selector!(#selector));
                    #(#key_fields)*

                    let mut data = array![];
                    #(#data_fields)*

                    let expected = Event { keys, data };
                    self.assert_only_event(contract, expected);
                }

                #assert_only_fn
            }
        })
        .collect();

    let maybe_pub = if impl_block.is_public {
        quote! { pub }
    } else {
        quote! {}
    };

    quote! {
        #[generate_trait]
        #maybe_pub impl #impl_name of #trait_name {
            #(#methods)*
        }
    }
}
