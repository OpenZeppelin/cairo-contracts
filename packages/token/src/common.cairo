pub mod erc2981;
