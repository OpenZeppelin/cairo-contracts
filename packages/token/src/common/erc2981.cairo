pub mod erc2981;
pub mod interface;

pub use erc2981::{ERC2981Component, DefaultConfig};
pub use interface::{IERC2981Dispatcher, IERC2981DispatcherTrait};
