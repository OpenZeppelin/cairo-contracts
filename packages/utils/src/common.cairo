// SPDX-License-Identifier: MIT
// OpenZeppelin Contracts for Cairo v0.15.1 (utils/common.cairo)

use core::traits::PartialOrd;

pub impl Felt252PartialOrd of PartialOrd<felt252> {
    #[inline(always)]
    fn le(lhs: felt252, rhs: felt252) -> bool {
        let lhs: u256 = lhs.into();
        lhs <= rhs.into()
    }
    #[inline(always)]
    fn ge(lhs: felt252, rhs: felt252) -> bool {
        let lhs: u256 = lhs.into();
        lhs >= rhs.into()
    }
    #[inline(always)]
    fn lt(lhs: felt252, rhs: felt252) -> bool {
        let lhs: u256 = lhs.into();
        lhs < rhs.into()
    }
    #[inline(always)]
    fn gt(lhs: felt252, rhs: felt252) -> bool {
        let lhs: u256 = lhs.into();
        lhs > rhs.into()
    }
}
