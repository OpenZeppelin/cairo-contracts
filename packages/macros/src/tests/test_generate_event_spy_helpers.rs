use cairo_lang_macro::{quote, TokenStream};
use insta::assert_snapshot;

use super::common::format_proc_macro_result;
use crate::generate_event_spy_helpers::definition::generate_event_spy_helpers_q2hsv6hnhh3ug as generate_event_spy_helpers;

#[test]
fn test_event_without_attrs() {
    let input = quote!(
        {
            pub impl MockContractSpyHelpers {
                event SimpleEvent(
                    field: u64
                );
            }
        }
    );
    assert_snapshot!(get_string_result(input));
}

#[test]
fn test_event_with_only() {
    let input = quote!(
        {
            pub impl MockContractSpyHelpers {
                #[only]
                event OnlyEvent(
                    field: u64
                );
            }
        }
    );
    assert_snapshot!(get_string_result(input));
}

#[test]
fn test_event_with_key() {
    let input = quote!(
        {
            pub impl ContrMockContractSpyHelpersactSpy {
                event KeyedEvent(
                    #[key]
                    user: ContractAddress,
                    value: u64
                );
            }
        }
    );
    assert_snapshot!(get_string_result(input));
}

#[test]
fn test_event_with_only_and_key() {
    let input = quote!(
        {
            pub impl MockContractSpyHelpers {
                #[only]
                event OnlyKeyedEvent(
                    #[key] user: ContractAddress,
                    value: u64
                );
            }
        }
    );
    assert_snapshot!(get_string_result(input));
}

#[test]
fn test_event_with_invalid_non_default_attr() {
    let input = quote!(
        {
            pub impl MockContractSpyHelpers {
                #[non_default]
                event InvalidEvent(
                    field: u64
                );
            }
        }
    );
    assert_snapshot!(get_string_result(input));
}

#[test]
fn test_impl_without_pub() {
    let input = quote!(
        {
            impl MockContractSpyHelpers {
                event MockEvent(
                    field: u64
                );
            }
        }
    );
    assert_snapshot!(get_string_result(input));
}

#[test]
fn test_empty_impl() {
    let input = quote!({
        impl MockContractSpyHelpers {}
    });
    assert_snapshot!(get_string_result(input));
}

#[test]
fn test_multiple_events() {
    let input = quote!(
        {
            impl MockContractSpyHelpers {
                event MockEvent(
                    field: u64
                );

                #[only]
                event MockEvent2(
                    field: u64
                );
            }
        }
    );
    assert_snapshot!(get_string_result(input));
}

#[test]
fn test_empty_event() {
    let input = quote!(
        {
            impl MockContractSpyHelpers {
                #[only]
                event EmptyEvent();
            }
        }
    );
    assert_snapshot!(get_string_result(input));
}

fn get_string_result(args_stream: TokenStream) -> String {
    let raw_result = generate_event_spy_helpers(args_stream);
    format_proc_macro_result(raw_result)
}
