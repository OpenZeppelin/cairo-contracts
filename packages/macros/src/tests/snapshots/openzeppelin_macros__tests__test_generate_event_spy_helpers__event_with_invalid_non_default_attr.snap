---
source: src/tests/test_generate_event_spy_helpers.rs
expression: get_string_result(input)
snapshot_kind: text
---
TokenStream:

None

Diagnostics:

====
Error: Parsing Error: Error { input: "#[non_default]event InvalidEvent(field:u64);}}", code: Char }====

AuxData:

None
